File: leetcode/112/src/P112.java
/**
 * Created by qwordy on 2/26/16.
 * 112. Path Sum
 */

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public boolean hasPathSum(TreeNode root, int sum) {
		return root != null && search(root, sum, 0);
	}

	private boolean search(TreeNode node, int sum, int crtSum) {
		crtSum += node.val;
		return node.left == null && node.right == null && crtSum == sum || node.left != null && search(node.left, sum, crtSum) || node.right != null && search(node.right, sum, crtSum);
	}
}

public class P112 {
}

File: leetcode/112/P112.rb
# 112. Path Sum
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} sum
# @return {Boolean}
def has_path_sum(root, sum)
	(root && search(root, sum, 0)) ? true : false
end

def search node, sum, crtSum
	crtSum += node.val
	!node.left && !node.right && crtSum == sum || node.left && search(node.left, sum, crtSum) || node.right && search(node.right, sum, crtSum)
end

